Inetd Enhancements

ritzow opened the pull request:
Things we did not get to implement:

"Prefork: Support pre-forking multiple children and keeping them alive for multiple invocations."
"Improve the logging and make logging modes configurable on a per-service basis."
"Improve the logging and make logging modes configurable on a per-service basis."
"Add a separate way to turn services on and off, so they can be defined statically (such as in /usr/share) and turned on and off from /etc."
"Allow non-privileged users to add/remove/change their own services using a separate utility."
"Integrate with the new blacklist daemon." (It looks like this change was implemented already in the Makefile)
"Configuration compatibility for systemd socket activations"

Here are some notes about our changes in no particular order:

All new functionality should be explained by the updated manpage.

The manpage has been refactored a bit: A new section “Directives” has been added and the information about default hostnames and IPsec directives has been moved there, and the new file include directive information is also there.

getconfigent has the most major changes. A newline is no longer read immediately, but is called only by a “goto more” (inside an if(false) block). This allows multiple definitions or directives to exist on a single line for anything that doesn’t terminate using a newline. This means a key-values service definition can be followed by another key-values service definition, a positional definition, or an ipsec, hostname, or .include directive on the same line.

memset is no longer used explicitly to clear the servtab structure, a function init_servtab() is used instead, which uses a C struct initializer.

The servtab se_group field is its own allocation now, and not just a pointer into the user:group string.

Refactored some stuff out of getconfigent to separate functions for use by parse_v2.c. These functions in inetd.c are named with the form parse_*()

parse_v2.c only has code for parsing a key-values service definition into a provided servtab. It should not have anything that affects global state other than line and line_number.

Some function prototypes, structures, and #defines have been moved from inetd.c to inetd.h.

The function config_root replaces config as the function called on a config file load/reload. The code removed from the end of config(void) is now called in config_root, so it is not run on each recursive config call.

setconfig(void) was removed and its code added into config_root because that is the only place it is called, and redundant checks for non-null globals were removed because they are always freed by endconfig. The fseek code was also removed because the config files are always closed by endconfig.

Rate limiting code was updated to add a per-service per-IP rate limiting form. Some of that code was refactored out of other places into functions with names in the form rl_*()

We have not added any of the license or version information to the new files parse_v2.c, parse_v2.h, and inetd.h and we have not updated the license or version info for inetd.c.


Security related:

The behavior when reading invalid IPsec strings has changed. Inetd no longer exits, it quits reading the current config file instead. Could this impact program security?

We have not checked for memory leaks. Solomon tried to use dmalloc without success. getconfigent seemed to have a memory leak at each “goto more”. It seems like inetd has never free’d allocated strings when throwing away erroneous service definitions during parsing (i.e. when “goto more” is called when parsing fields). OpenBSD’s version calls freeconfig on “goto more” (https://github.com/openbsd/src/blob/c5eae130d6c937080c3d30d124e8c8b86db7d625/usr.sbin/inetd/inetd.c#L1049) but NetBSD only calls it when service definitions are no longer needed. This has been fixed. freeconfig is called immediately before any “goto more”. There shouldn’t be any time when a servtab is in an invalid state where freeconfig would break.

ritzow committed:
Inetd Enhancements

Things we did not get to implement:

"Prefork: Support pre-forking multiple children and keeping them alive for multiple invocations."
"Improve the logging and make logging modes configurable on a per-service basis."
"Improve the logging and make logging modes configurable on a per-service basis."
"Add a separate way to turn services on and off, so they can be defined statically (such as in /usr/share) and turned on and off from /etc."
"Allow non-privileged users to add/remove/change their own services using a separate utility."
"Integrate with the new blacklist daemon." (It looks like this change was implemented already in the Makefile)
"Configuration compatibility for systemd socket activations"

Here are some notes about our changes in no particular order:

All new functionality should be explained by the updated manpage.

The manpage has been refactored a bit: A new section “Directives” has been added and the information about default hostnames and IPsec directives has been moved there, and the new file include directive information is also there.

getconfigent has the most major changes. A newline is no longer read immediately, but is called only by a “goto more” (inside an if(false) block). This allows multiple definitions or directives to exist on a single line for anything that doesn’t terminate using a newline. This means a key-values service definition can be followed by another key-values service definition, a positional definition, or an ipsec, hostname, or .include directive on the same line.

memset is no longer used explicitly to clear the servtab structure, a function init_servtab() is used instead, which uses a C struct initializer.

The servtab se_group field is its own allocation now, and not just a pointer into the user:group string.

Refactored some stuff out of getconfigent to separate functions for use by parse_v2.c. These functions in inetd.c are named with the form parse_*()

parse_v2.c only has code for parsing a key-values service definition into a provided servtab. It should not have anything that affects global state other than line and line_number.

Some function prototypes, structures, and #defines have been moved from inetd.c to inetd.h.

The function config_root replaces config as the function called on a config file load/reload. The code removed from the end of config(void) is now called in config_root, so it is not run on each recursive config call.

setconfig(void) was removed and its code added into config_root because that is the only place it is called, and redundant checks for non-null globals were removed because they are always freed by endconfig. The fseek code was also removed because the config files are always closed by endconfig.

Rate limiting code was updated to add a per-service per-IP rate limiting form. Some of that code was refactored out of other places into functions with names in the form rl_*()

We have not added any of the license or version information to the new files parse_v2.c, parse_v2.h, and inetd.h and we have not updated the license or version info for inetd.c.


Security related:

The behavior when reading invalid IPsec strings has changed. Inetd no longer exits, it quits reading the current config file instead. Could this impact program security?

We have not checked for memory leaks. Solomon tried to use dmalloc without success. getconfigent seemed to have a memory leak at each “goto more”. It seems like inetd has never free’d allocated strings when throwing away erroneous service definitions during parsing (i.e. when “goto more” is called when parsing fields). OpenBSD’s version calls freeconfig on “goto more” (https://github.com/openbsd/src/blob/c5eae130d6c937080c3d30d124e8c8b86db7d625/usr.sbin/inetd/inetd.c#L1049) but NetBSD only calls it when service definitions are no longer needed. This has been fixed. freeconfig is called immediately before any “goto more”. There shouldn’t be any time when a servtab is in an invalid state where freeconfig would break.

zoulasc commented on usr.sbin/inetd/inetd.h:
```diff
@@ -0,0 +1,128 @@
+#define MUX_TYPE	1
+#define MUXPLUS_TYPE	2
+#define FAITH_TYPE	3
+#define ISMUX(sep)	(((sep)->se_type == MUX_TYPE) || \
```
I would re-use ISMUXPLUS() here.

zoulasc commented on usr.sbin/inetd/inetd.h:
```diff
@@ -0,0 +1,128 @@
+		struct	sockaddr_in6 se_ctrladdr_in6; /* in6 is used by bind()/getaddrinfo */
+		struct	sockaddr_un se_ctrladdr_un;
+	};			/* bound address */
+	int	se_ctrladdr_size;
```
Perhaps sizes and counts should be size_t

sauclovian replied:
given that they're sockaddr sizes, socklen_t might be more direct useful

zoulasc replied:
ok!

zoulasc commented on usr.sbin/inetd/inetd.h:
```diff
@@ -0,0 +1,128 @@
+			 ((sep)->se_type == MUXPLUS_TYPE))
+#define ISMUXPLUS(sep)	((sep)->se_type == MUXPLUS_TYPE)
+
+#define	A_CNT(a)	(sizeof (a) / sizeof (a[0]))
```
NetBSD has __arraycount() for this.

ritzow replied:
This was part of the existing code which we just moved over to inetd.h but it would be good to update it now. I'll do that.

zoulasc commented on usr.sbin/inetd/inetd.h:
```diff
@@ -0,0 +1,128 @@
+#include "ipsec.h"
+#endif
+
+#define NORM_TYPE	0
```
could also use an enum

ritzow replied:
This was just existing inetd code. It's probably somewhat simple to update if you think this would be a good opportunity to do that.

zoulasc replied:
I think modernizing the code is good, let's do it.

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1067 @@
+	size_t line_number_start = line_number;
+	invoke_result result;
+
+	if (debug) {
```
Perhaps use a DPRINTF() macro to allow compiling out the debugging code.

ritzow replied:
The debugging code is enabled using the -d flag at runtime. Wouldn't that sort of negate the need for a macro? I think a DPRINTF macro would still be helpful and allow disabling of debugging for someone compiling the program who doesn't want it, but its purpose is mainly to check that a config is loaded correctly.

zoulasc replied:
I was thinking that a macro is less clutter, and can be compiled away if needed. So it has dual benefit.

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1067 @@
+	invoke_result result;
+
+	if (debug) {
+		fprintf(stderr, "Found key-values definition line %zu: '%s'\n", 
```
use warnx(3) instead so that the program name gets printed; also perhaps always prefix syntax errors with %s, %zu: (file and line) for consistency.

ritzow replied:
I'm sort of wondering what type of printouts the -d flag should enable. Before these changes, the things printed were adding, removing, or updating a service definition, and any runtime socket stuff. We added a bunch of debug printouts relating to config parsing like include directives and encountering a key-values service definition. 

I'm wondering if I should add a printout for encountering a "positional" service definition as well, or remove this one relating to key-values parsing. Syntax errors are also only written to the system log and not to stderr with the -d flag (although it could be changed to print to both). It does seem important to print the include directive events at least.

Basically, how "clean" do we want the -d output to be?

zoulasc replied:
I think that syntax errors should be reported consistently in syslog (because with -HUP you can be re-parsing while daemonized and stderr is not available), regardless of -d. I.e. the same error messages should be printed to syslog with or without -d. I think it makes sense for "positional" service definitions to behave like the traditional ones. If the key-value debugging flags are useful to determine what is causing the syntax errors, they should stay. You could turn them on with a higher debugging level (-dd) or having a special debugging flag -d parse etc, but I think both are overkill and I would leave it alone (since parsing does not happen all the time).

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1067 @@
+				if (debug) {
+					fprintf(stderr, "Invalid definition, ignoring\n");
+				}
+				syslog(LOG_ERR, CONF_ERROR_FMT "Ignoring invalid definition.",
```
Why doesn't this syslog use the macro?

ritzow replied:
It uses line_number_start so the error will point to the first line of the service definition instead of the last line. Not sure if I should make another macro that leaves the line number as a variadic parameter for use in this scenario.

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1067 @@
+		case INVOKE_FINISH:
+			/* Found a semicolon, do final checks and defaults and return. */
+			/* Skip whitespace after semicolon to end of line. */
+			while (**cpp == ' ' || **cpp == '\t') {
```
isspace(3)

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1067 @@
+ * Return true on success, false on failure. 
+ */
+static bool 
+fill_default_values(struct servtab * sep) {
```
Whitespace inconsistency: we use "type *var" not "type* var" or "type * var". There are others in this file.

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1067 @@
+	if (sep->se_policy == NULL) {
+		/* Set to default global policy */
+		sep->se_policy = policy;
+	} else if (strlen(sep->se_policy) == 0) {
```
*sep->se_policy == '\0'

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1067 @@
+
+static void
+try_infer_socktype(struct servtab * sep) {
+	if (sep->se_socktype == SE_SOCKTYPE_UNINIT) {
```
Use early returns style;
if (sep->se_socktype != SE_SOCKTYPE_UNINIT || sep->se_proto == NULL)
    return;

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1067 @@
+	if (parse_server(sep, "internal") != 0) {
+		ENI("exec");
+		return false;
+	} else if (wait_prev != SE_WAIT_UNINIT && wait_prev != sep->se_wait) {
```
Early returns style in the whole compilation unit: No point in using else, if you return. Early returns is good style because it "compartmentalizes" a function. You can forget what you just read because it has been taken care of, and you can focus on the rest of the function, instead of needing to "keep state".

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1067 @@
+	if (strcmp("tcp", sep->se_proto) == 0
+		|| strcmp("udp", sep->se_proto) == 0
+		|| strcmp("rpc/tcp", sep->se_proto) == 0
+		|| strcmp("rpc/udp", sep->se_proto) == 0) {
```
Same here. If you did not find something interesting return sooner. Also avoids code staircases.

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1067 @@
+	int line_start = line_number;
+	
+	for (;;) {
+		while (isblank((int)*cp))
```
cast to unsigned char. valid values for ctype macros are 0..255 and -1. A negative char cast to int has undefined behavior.

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1067 @@
+				return false;
+			}
+
+		} else break;
```
early returns again; if you "continue;" in the if part then you don't need an else.

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1067 @@
+static key_handler_func 
+get_handler(char * name) {
+	/* Call function to handle option parsing. */
+	for(unsigned int i = 0; i < A_CNT(key_handlers); i++) {
```
size_t

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1067 @@
+	while (*cp && quote) {
+		if (*cp == quote) {
+				quote = '\0';
+				memmove(cp, cp+1, strlen(cp)); 
```
again continue here..

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1067 @@
+			cp++;
+			switch (*cp) {
+			case 'x':
+				sscanf(cp+1, "%2hhx%n", start, &count_read);
```
sscanf and %n are evil. use custom hex parsing.

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1067 @@
+				}
+				memmove(cp, cp + count_read + 1, strlen(cp));
+				break;
+			case '\\':
```
no point in repeating the memmove so many times. It can be done once at the bottom of the switch if you make the hex case to continue instead of break.

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1067 @@
+	 * past the end of the value. 
+	 */
+	char * start = cp;
+	while (!isblank((int)*cp) && *cp != '#' && 
```
cast again.

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1067 @@
+	val_parse_info info;
+
+	/* Skip any whitespace if it exists, otherwise do nothing */
+	if (!skip_whitespace(&cp)) {
```
you could use strsep(3) or stresep(3) to simplify parsing.

zoulasc replied:
Forget about it :-)

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1067 @@
+
+		
+	/* alphabetical or underscore allowed in name */
+	while (isalpha((int)*cp) || *cp == '_') {
```
all ctype(3) casts need fixing.

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1067 @@
+	/* TODO need to figure out difference between v1 and v2 */
+	char* tail;
+	errno = 0;
+	long count = strtol(count_str, &tail, 10);
```
we have strtoi(3) which is less clumsy.

zoulasc commented on usr.sbin/inetd/parse_v2.h:
```diff
@@ -0,0 +1,10 @@
+#include "inetd.h"
+
+typedef enum parse_v2_result {V2_SUCCESS, V2_SKIP, V2_ERROR} parse_v2_result;
```
Looking at all the parsing code, I wonder if it would not have been simpler to use lex and yacc...

ritzow replied:
Yeah Alex looked into this a bit and created a simple lex/yacc for some of the syntax. Three of us took a compiler class last quarter and realized it would probably have been simpler. Even a hand-done scanner/parser might be simpler. Not sure what the best solution would be at this point. The advantage of the code we have is it just marches across the text without requiring a special yacc stack or anything else. Each key is looked up and just becomes a function pointer, and each key handler reads values one by one at its own discretion instead of being driven by the parser, so the updating of the servtab is somewhat straightforward, and it's pretty efficient. But I don't know how much that's worth the complexity or verbosity of the code we have.

zoulasc replied:
Yes, learning lex and yacc is useful for things like this and are good tools to know. I am not suggesting to re-do the whole thing, but something to keep in mind...

zoulasc commented on usr.sbin/inetd/inetd.h:
```diff
@@ -0,0 +1,128 @@
+#define ILV(prio, msg, args...) syslog(prio, CONF_ERROR_FMT msg ".", \
+	CONFIG, line_number, args)
+
+#define WRN0(msg) IL0(LOG_WARNING, msg)
```
These can be done better with ```__VA_OPT__(,)```

ritzow replied:
This is perfect, I'll remove ERR0 and WRN0 and use this with ERR and WRN instead.

zoulasc commented on usr.sbin/inetd/inetd.h:
```diff
@@ -0,0 +1,128 @@
+extern char *policy;
+
+/* "Unspecified" indicator value for servtabs (mainly used by v2 syntax) */
+#define SE_SERVICE_MAX_UNINIT -1
```
What is the advantage of having 4 definitions of -1 to mean the same thing for the same type of variable? I would think SE_UNINIT would suffice :-)

ritzow replied:
Done 👍 (I'm changing it to SERVTAB_UNSPEC_VAL which makes more sense than "uninit", since its specifically an initializer. Not sure why I used uninit before.)

ritzow committed:
Partial changes from code review

ritzow committed:
More pointer style updates

ritzow committed:
Replace strtol with strtoi

ritzow committed:
Miscellaneous changes

Change se_ctrladdr_size from int to socklen_t. Create a DPRINTF macro for debug code and update code to use it.

ritzow committed:
Fix parse_quotes

ritzow committed:
Reimplement hexcode parsing in parse_quotes and make other fixes

ritzow committed:
Config parsing related debug prints should print file/line info

ritzow commented:
Basically all suggestions so far have now been implemented with the exception of using warnx, since that seemed more about syntax errors and warnings which are just syslog'd as usual. I added the DPRINTF macro, and a DPRINTCONF macro that looks similar to a syntax error but for debug prints instead of errors, so the debug output should look pretty uniform.

zoulasc commented on usr.sbin/inetd/inetd.h:
```diff
@@ -0,0 +1,136 @@
+	socklen_t	se_ctrladdr_size;
+	int	se_service_max;		/* max # of instances of this service */
+	int	se_count;		/* number of instances of this service started since se_time */
+	int	se_ip_max;  		/* max # of instances of this service per ip per minute */
```
should those counts be size_t? Can they be negative? If not then they should probably be size_t

ritzow replied:
I think se_service_max and se_ip_max can be -1 so that any repeat definitions of them in a key-values service definition can be caught, but that isn't important. We can change it to detect if it is already greater than zero and not worry about if someone sets it to zero and then another value. I put in the manual page that not all errors are necessarily detected.

zoulasc replied:
Or you can make the invalid/unititialized value size_t too:
#define UNINIT ((size_t)~0)

ritzow replied:
Sounds perfect 👍.

zoulasc commented on usr.sbin/inetd/inetd.h:
```diff
@@ -0,0 +1,136 @@
+	int	se_ip_max;  		/* max # of instances of this service per ip per minute */
+	struct se_ip_list_node{
+		char address[14];
+		int count;		/* number of instances of this service started from
```
Perhaps this one too.

ritzow replied:
I just noticed that char address[14] is 14 bytes for some reason. I'll ask Alex about this code, it seems like address might be better as a union of in_addr and in6_addr. It should be at least 16 bytes if it is what I assume it is. We can probably change count too.

zoulasc commented on usr.sbin/inetd/inetd.h:
```diff
@@ -0,0 +1,136 @@
+
+/* From inetd.c */
+int 	parse_protocol(struct servtab *);
+int 	parse_wait(struct servtab *, int wait);
```
no names in prototype decls (int wait -> int).

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1110 @@
+static hresult	filter_handler(struct servtab *, vlist);
+static hresult	group_handler(struct servtab *, vlist);
+static hresult	service_max_handler(struct servtab *, vlist);
+static hresult  ip_max_handler(struct servtab *, vlist);
```
looks misindented.

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1110 @@
+					*cpp = nextline(fconfig);
+				}
+				return V2_SUCCESS;
+			} else {
```
Early return; no need for else here.

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1110 @@
+static key_handler_func 
+get_handler(char *name) {
+	/* Call function to handle option parsing. */
+	for(size_t i = 0; i < __arraycount(key_handlers); i++) {
```
space after keyword (for)

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1110 @@
+	char *cp = *cpp;
+	char quote = *cp;
+
+	memmove(cp, cp+1, strlen(cp+1)+1); 	
```
spaces around operators (+)

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1110 @@
+	while (*cp && quote) {
+		if (*cp == quote) {
+			quote = '\0';
+			memmove(cp, cp+1, strlen(cp+1)+1);
```
I think this memmove pattern is used too frequently and could be made into an inline function:
```
static inline void 
strmove(char *buf, size_t off) { 
            memcpy(buf, buf + off, strlen(buf + off) + 1);
}

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1110 @@
+}
+
+char hex_to_bits(char in) {
+	switch(in) {
```
```
unsigned char c = (unsigned char)in;
if (!isxdigit(d))
      return -1;
// ascii dependency
return isdigit(d) ? d - '0' : tolower(d) - 'a' + 10;
```
or
```
static const char xdigits[] = "0123456789abcdef";
char *p = strchr(xdigits, tolower(c));
return p == NULL ? -1 : p - xdigits;
```
or
```
        switch (in) {
        case '0'...'9':
                return in - '0';
        case 'a'...'f':
                return in - 'a' + 10;
        case 'A'...'F':
                return in - 'A' + 10;
        default:
                return -1;
        }
```

ritzow replied:
I saw that case range operator when I was doing some google searching and wasn't sure if we were able to use that GCC feature. I'm obsessed with switch statements so I might just go with the third one. Concise and efficient (although I suppose you never know how the jump table might turn out, could just be a bunch of conditionals).

zoulasc replied:
I thought it would have made it in the c standard by now. anyway, both clang and gcc support it... and gcc does not even turn it off with gcc -ansi -std=c89, weird.

zoulasc commented on usr.sbin/inetd/parse_v2.h:
```diff
@@ -0,0 +1,10 @@
+#include "inetd.h"
+
+typedef enum parse_v2_result {V2_SUCCESS, V2_SKIP, V2_ERROR} parse_v2_result;
```
Perhaps fold parse_v2.h this into inetd.h?

ritzow replied:
We had it in inetd.h up until a couple days ago when I separated them because I thought it might be more correct to have a separate parse_v2.h :D. It does feel a little silly to have such a small header. I'll merge them back.

zoulasc commented on usr.sbin/inetd/inetd.h:
```diff
@@ -0,0 +1,136 @@
+#ifndef _INETD_H
```
Copy the copyright from inetd.c to inetd.h since the code was lifted from there.

zoulasc commented on usr.sbin/inetd/parse_v2.c:
```diff
@@ -0,0 +1,1110 @@
+#include <ctype.h>
```
Add copyright, RCSID?

zoulasc reviewed (commented):
Getting real close. I added a few more!

ritzow committed:
Implement more code review changes

Style conformance
Change servtab field types to size_t and make unspecified value SIZE_MAX
parse_quotes redundancy reduction
Merge parse_v2.h into inetd.h

Note: Changes to make all 'int to size_t' fields work properly may not be complete.

ritzow committed:
Change memcpy to memmove in strmove

ritzow committed:
Fixes for integer parsing and rate limiting

Also update manpage for IP rate limiting

ritzow commented:
Our IP rate limiting seems a little wonky so I'm attempting to do a fix for that.

ritzow committed:
More code review changes and fix IP rate limiting

IP rate limiting should work properly now.
Fix segfault in IP rate limiting code.
Add clear_ip_list to reduce code redundancy.
Remove extra copyright comment from parse_v2.c.
Update manpage on IP rate limiting, add comment about "max"/"service_max" value off-by-one to BUGS section.
Add struct "struct sockaddr_storage se_ctrladdr_storage;" to sockaddr union so the socket address is properly zero'd by the struct initializer in init_servtab.

ritzow committed:
Delete parse_v2.h

ritzow commented:
The IP rate limiting should be fixed now I think. If an IP address connects too many times it won't shut down the entire service now, it just prevents forking the process for that connection. I haven't tested the changes with a "wait" type service or a UDP service though.

zoulasc commented:
Great, thanks!

ritzow commented:
Looks like I need to fix UDP IP rate limiting. getpeername only works for connected sockets so I need to use recvmsg to get the source address of the packet.

Edit: I've fixed it so now SOCK_STREAM services use getpeername and SOCK_DGRAM services use recvmsg and peek at the packet to get the remote IP. For any other type it just fails the rate limiting with a syslogged error.

I found that even the regular service_max (previous just "max") seems to fail to prevent a "nowait" service from spawning. However I've only tried it with huntd so far, which is actually a wait service.

ritzow committed:
Fix getting remote address for IP rate limiting with UDP

ritzow committed:
More rate limiting fixes

Fixed some stuff I might have broken during previous edits.
Made sep->se_family into sa_family_t type.

ritzow commented on usr.sbin/inetd/inetd.c:
```diff
@@ -2761,25 +2752,29 @@ rl_process(struct servtab *sep, int ctrl)
+			    "This service has se_ip_max %zu and ip_count %zu",
 			    sep->se_ip_max, node->count);
-			if (node->count == sep->se_ip_max) {
+			if (node->count >= sep->se_ip_max) {
```
Oops! I changed this previously I think, just reverted it.

ritzow commented on usr.sbin/inetd/inetd.c:
```diff
@@ -2748,8 +2737,10 @@ rl_process(struct servtab *sep, int ctrl)
-			sizeof(hbuf), NULL, 0, NI_NUMERICHOST)) {
+
+        	if (getnameinfo((struct sockaddr *)&addr, 
+		    ((struct sockaddr *)&addr)->sa_len, hbuf, 
```
I guess on BSD sockaddr has a non-standard sa_len field (I changed it to something else but changed it back to what Alex and Gabe had). I think this should work with recvmsg, recvmsg doesn't set the namelen field in the msghdr.

ritzow commented on usr.sbin/inetd/inetd.c:
```diff
@@ -2761,25 +2752,29 @@ rl_process(struct servtab *sep, int ctrl)
-                        		    sep->se_proto,
-                        		    sep->se_ip_max,
-					    node->address);
+					if (debug) {
+						/* Only log during debug to prevent
+						DoS attack writing to system log */
+						syslog(LOG_ERR,
+                        		   	    "%s/%s max ip spawn rate (%zu in " 
+					   	    TOSTRING(CNT_INTVL) " seconds) for "
+					   	    "%." TOSTRING(NI_MAXHOST) "s "
+                        		   	    "exceeded; service not started.",
+                        		   	    sep->se_service,
+                        		   	    sep->se_proto,
+                        		   	    sep->se_ip_max,
+					   	    node->address);
+					}
```
Since this IP rate limiting isn't perfect, and every incoming packet and connection is processed, this could lead to a lot of syslog garbage if there was a DoS attack from multiple IPs. Don't want any chance of that.

ritzow commented on usr.sbin/inetd/inetd.h:
```diff
@@ -122,7 +122,7 @@ struct	servtab {
 	char	*se_service;		/* name of service */
 	int	se_socktype;		/* type of socket to use */
-	int	se_family;		/* address family */
+	sa_family_t	se_family;	/* address family */
```
sa_family_t might be smaller (an unsigned short) than the previous int value but I don't think anything relies on the exact struct layout and we already changed se_count and other fields.

ritzow commented:
Rate limiting seems to generally work in wait and nowait modes for TCP and UDP. It's a bit hard to test considering there aren't really any UDP nowait programs that are easily testable, but I don't think there are any times when it doesn't work, outside of the general inaccuracy of using a single timestamp to mark the beginning of the service and ip count interval.

zoulasc commented:
You are right, if you want to test this, I think the best way is to write your own service. It should only be a few lines long.

ritzow committed:
Test and fix ip_max rate limiting with nowait dgram service

Replace gettimeofday with clock_gettime for rate limiting, more debug improvements
Standardize service names in debug/syslog printouts, IP rate limiting fix, update manpage

ritzow commented on usr.sbin/inetd/inetd.h:
```diff
@@ -213,4 +213,8 @@ parse_v2_result	parse_syntax_v2(struct servtab *, char **);
 
 #define SERVTAB_COUNT_MAX (SIZE_MAX - (size_t)1)
 
+/* Standard logging and debug print format for a servtab */
+#define SERV_FMT "%s/%s"
+#define SERV_PARAMS(sep) sep->se_service,sep->se_proto
+
```
These formats are now used any time a service name is printed.  SERV_FMT in the printf format string, SERV_PARAMS is the parameters.

ritzow committed:
Another rate limiting fix

se_count is now only incremented on successfully passing all rate limiting checks. ip_max rate limiting now consumes a datagram whether nor not the service is wait or nowait.

ritzow commented on usr.sbin/inetd/inetd.c:
```diff
@@ -2674,7 +2674,8 @@ rl_process(struct servtab *sep, int ctrl)
 	    sep->se_service_max, sep->se_count);
 
-	if (sep->se_count++ == 0) {
+	/* se_count is incremented when rl_process returns 0 */
+	if (sep->se_count == 0) {
```
Only increment se_count when successfully passing rate limiting checks, otherwise this is just "packet" rate limiting and won't work very well next to ip rate limiting.

ritzow committed:
Update manpage

ritzow committed:
Fix max/service_max always allowing one service to start and allow one less than specified

ritzow commented on usr.sbin/inetd/inetd.c:
```diff
@@ -2679,7 +2679,9 @@ rl_process(struct servtab *sep, int ctrl)
 		sep->se_time = now;
 		istimevalid = true;
-	} else if (sep->se_count >= sep->se_service_max) {
+	}
+	
+	if (sep->se_count >= sep->se_service_max) {
```
This is a behavior change for inetd. Services that specified 0 for max and expected 1 service to start will no longer start. Services that specified N for max and expected (N-1) services to start will allow one more service than expected.

ritzow replied:
This is a behavior change for inetd. Services that specified 0 for max and expected 1 service to start will no longer start. Services that specified N for max and expected (N-1) services to start will allow one more service than expected. Should I keep this or revert it? It seems like it would be extremely odd behavior for someone to rely on, but you never know.

ritzow commented:
The IP rate limiting should be good to go now. I wrote a little udp client and server to test nowait dgram services and fixed a couple bugs with that. There's probably some undefined behavior with services that aren't specifically TCP or UDP, although it's only discussed as IP rate limiting in the manual page so I think that should be okay. I also polished up a lot of debug print messages so they should be much easier to read now. Have you had a chance to test out the new rate limiting, Christos?

zoulasc commented:
No I have not yet. I am wondering if it would be easy/possible to tedt inetd using ATF or even if it is worth to make it use rump (probably not)

ritzow commented:
Alex and Gabe created ATF tests for inetd, but they were only some regression tests for the config files and we ended up not including them since they wouldn't work well in a regular installation without changes. I don't really know what would work best. 

I should also double check my understanding of nowait dgram services. I determined that they are implemented by reading a single dgram message from the passed in socket and then doing their own thing. Is that correct?

Here's my code for the nowait UDP client and server:

```C
#include <sys/socket.h>
#include <netdb.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>

#define UDP 17

void main(size_t argc, char * argv[argc + 1]) {
	char buffer[65535];
	char name[NI_MAXHOST];
	int udp = socket(PF_INET, SOCK_DGRAM, UDP);

	if(argc != 4) {
		printf("Invalid arguments\n");
	} else {
		struct addrinfo hints = {
			.ai_socktype = SOCK_DGRAM,
			.ai_protocol = UDP
		};
		struct addrinfo * res;
		int error;

		if(error = getaddrinfo(argv[1], argv[2], &hints, &res)) {
			fprintf(stdout, "Error: %s\n", gai_strerror(error));
		} else {
			sendto(udp, argv[3], strlen(argv[3]), 0, res->ai_addr, res->ai_addrlen);
			printf("Waiting to receive message...\n");
			ssize_t count = recv(udp, buffer, sizeof(buffer), 0);
			printf("Received message: \"%s\"\n", buffer);
		}
	}
}
```


```C
#include <sys/socket.h>
#include <netdb.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <syslog.h>

#define UDP 17	

void main() {
	openlog("udpserver", LOG_PID | LOG_NOWAIT, LOG_DAEMON);
	char buffer[65535];
	char name[NI_MAXHOST];
	socklen_t source_size;
	struct sockaddr_storage addr;
	struct iovec store = {
		.iov_base = &buffer,
		.iov_len = sizeof(buffer)
	};
	struct msghdr header = {
		.msg_name = &addr,
		.msg_namelen = sizeof(struct sockaddr_storage),
		.msg_iov = &store,
		.msg_iovlen = 1
		/* scatter/gather and control info is null */
	};
	int count;
	/* Peek so service can still get the packet */
	count = recvmsg(0, &header, 0);
	if (count == -1) {
		syslog(LOG_ERR,
			"failed to get dgram source address: %s",
			strerror(errno));
			return;
	}
	int error;
	if(error = getnameinfo((struct sockaddr*)&addr, addr.ss_len, name, NI_MAXHOST,
		NULL, 0, NI_NUMERICHOST)) {
		syslog(LOG_ERR, "Error: %s\n", gai_strerror(error));
		return;
	} else {
		syslog(LOG_ERR, "Received message from %s\n", name);
	}
	sendto(0, buffer, count, 0, (struct sockaddr*)header.msg_name, ((struct sockaddr*)header.msg_name)->sa_len);
	syslog(LOG_ERR, "done");
}
```

ritzow committed:
Some more rate limiting fixes/normalizations

A couple more things I missed for some reason. rl_reset and rl_add should now start se_count and node->count at 0 instead of 1. Rate limits of 0 should now always prevent any services from spawning.

ritzow committed:
Refactor rl_process 

Factor code into separate functions
Remove last remaining state-machine-like code
Change sep->se_time into time_t (seconds only) and modify rl_time()

ritzow commented:
Sorry, moved a bunch of code around in rate limiting. It should be a bit easier to read the rate limiting flow now.

ritzow committed:
Add missing IPSEC #ifdef, surround all debug prints with #ifdef

ritzow committed:
Remove unnecessary #ifdef I just added, oops

Commit:
Implement initial IP rate limiting ATF test for inetd

ritzow commented:
@zoulasc I uploaded a small IP rate limiting test case to the src/tests/usr.sbin/inetd directory. It compiles the little UDP echo server using gcc then executes inetd with a single service config file which uses the compiled server with ip_max of 3. It then attempts to query the echo server 4 times, and expects the final one to fail with a timeout. I could add a lot of other test cases to this to test different aspects. It's a bit janky having to call gcc from the test case and calling sleep(1) to "avoid" a race condition of sending a packet before inetd loads the service. It also expects inetd to not be running already and for port 5432 to be open.

zoulasc commented:
Thanks, I will take a look tomorrow or Tuesday...

Commit:
Add a few more rate limiting tests

ritzow committed:
Implement stream/wait and stream/nowait rl tests, format test code

ritzow commented:
Christos, have you had a chance to review the rate limiting test code I added yet? What can I do to make the code easier to review? Are there other tests I should implement? I could potentially write a couple regression tests for existing inetd behavior.

zoulasc commented:
No, sorry I have not. I will do that tomorrow!

zoulasc commented:
You can build the helper programs as part of the test suite; I think they should use standard err(3) and warn(3) instead of syslog. They will be installed from the test directory and then you can invoke them from an atf test shell script together with your own version of inetd that has a private configuration file. Once the test is done, it can kill everything it started. Is that possible? The tests in /usr/src/lib/libc/net have some tests that use helper programs and shell scripts. They also use rump hijack to hijack system calls to make daemons behave differently. I am not sure that you need that part, if you can run your own copy of inetd without interfering with the system one.

ritzow commented:
> /usr/src/lib/libc/net have some tests that use helper programs and shell scripts

This is very helpful!

I'll make sure to test it with system inetd also running.

ritzow commented:
The one thing that might interfere in terms of the inetd daemon are the localhost ports that are used (5432-5439). I can look into rump but it might be difficult to "mock" the whole network stack so used ports don't interfere with testing and vice versa. But maybe it doesn't require that much.

ritzow committed:
Minor inetd changes

ritzow committed:
Oops!

ritzow committed:
Update inetd test with code review changes

Don't use syslog in test_server. Compile test_server ahead of time.

ritzow committed:
Remove unused udpserver.c

zoulasc commented:
The comment about the helper programs and syslog() was incorrect. Yes the server that is running from inetd should use syslog.

ritzow committed:
Use syslog again for test_server

ritzow commented:
I changed the test_server back to using syslog. It doesn't seem like using rump will be possible since rump can't execute processes and inetd needs to be able to fork to function. The ones I implemented should work as long as the specific port range isn't in use.

zoulasc commented:
Great, what is left to do now?

ritzow commented:
I don't know of anything else to do as long as the couple small tests I wrote are sufficient. They're pretty simple tests and don't cover a whole lot but at least they confirm that the rate limiting prevents process spawns.

zoulasc commented:
Great, I guess that the next step is to patch my source tree, test it, and merge on CVS. Should I just add the standard NetBSD copyrights to the files? Who should be listed as authors?

ritzow commented:
Yeah, no special copyrights or anything, the standard ones are good. Here are the authors: James Browning <jamesbrowning137@gmail.com>, Gabe Coffland <GCoffland@gmail.com>, Alex Gavin <a_gavin@icloud.com>, Solomon Ritzow <solduck@gmail.com>. You can go ahead with next steps.

zoulasc commented:
Committed! Here are some suggestions to improve things:
- inetd.c is huge now. Let's move the ratelimit code to a new file ratelimit.c
- make the ratelimit list code use <sys/queue.h>
- perhaps move all the parsing code from inetd.c into a new parse.c

ritzow commented:
Awesome! Thanks for answering our questions and reviewing the code and working with me on these final changes. I can look into making those suggested changes. Might be a while though.

zoulasc commented:
Thank you all for making inetd better! It was a pleasure working with you.
